internal/network/connectivity_groups.go

```diff
@@ -142,116 +143,235 @@ func (g groupId) isLess(other groupId) bool {
-	groupList := createGroupList(groupCandidates)
-	ret := filterFullMeshGroups(groupList)
+func (c *connectivityGroupList) addOrMergeGroup(cg *connectivityGroup) {
+	foundGroup := c.findGroup(cg)
+	if foundGroup != nil {
+		foundGroup.mergeParticipants(cg)
+	} else {
+		c.addGroup(cg)
+	}
+}
 
-	// Sort by set size descending, which means the largest group first.
+func (c *connectivityGroupList) largestGroup() *connectivityGroup {
+	size := c.biggestGroupSize()
+	if size < 3 {
+		return nil
+	}
+	groupsBySize := c.groupsBySize[size]
+	var ret []*connectivityGroup
+	var numParticipatingHosts int
+	for _, cg := range groupsBySize {
+
+		// Take only the participants with the highest number of participating hosts
+		switch {
+		case cg.numParticipatingHosts() > numParticipatingHosts:
+			numParticipatingHosts = cg.numParticipatingHosts()
+			ret = nil
+			fallthrough
+		case cg.numParticipatingHosts() == numParticipatingHosts:
+			ret = append(ret, cg)
+		default:
+		}
+	}
 	sort.Slice(ret, func(i, j int) bool {
-		// If the sizes are equal then compare the contained elements of each set
 		return ret[i].id().isLess(ret[j].id())
 	})
+	return ret[0]
+}
+
+func (c *connectivityGroupList) biggestGroupSize() int {
+	ret := 0
+	for key := range c.groupsBySize {
+		ret = max(ret, key)
+	}
 	return ret
 }
 
+func (c *connectivityGroupList) reduce(size int) {
+	delete(c.groupsBySize, size)
+}
+
+func (c *connectivityGroupList) toList() (ret []*connectivityGroup) {
+	for _, l := range c.groupsBySize {
+		ret = append(ret, l...)
+	}
+	return ret
+}
+
+func (c *connectivityGroupList) setAndMergePrimaryGroups() []*connectivityGroup {
+	primaryGroups := c.toList()
+	for i, cg := range primaryGroups {
+		cg.setPrimaryGroup(len(primaryGroups), i)
+	}
+	return primaryGroups
+}
+
+// Group candidate is the connectivity view of a single host (me)
+type groupCandidate struct {
+	set *connectivitySet
+	me  int
+}
+
+func (m *majorityGroupCalculator) createFullMeshGroup(groupCandidates []*groupCandidate) *connectivityGroup {
+	groupList := newConnectivityGroupList()
+
+	// First iteration - gather the original groups
+	for _, candidate := range groupCandidates {
+
+		// Add the set of the current candidate to the pending list
+		groupList.addOrMergeGroup(m.toConnectivityGroup(candidate))
+	}
+
+	// The primary groups are used for intersection with group candidate which is not full mesh
+	primaryGroups := groupList.setAndMergePrimaryGroups()
+	for {
+		groupCandidate := groupList.largestGroup()
+		if groupCandidate != nil {
+			currentSize := groupCandidate.len()
+
+			if groupCandidate.isFullMesh() {
+				return groupCandidate
+			}
+
+			// Remove all groups with the current size, since we are now looking for smaller size groups
+			groupList.reduce(currentSize)
+
+			// We now want to intersect with groups not smaller than the biggest group(s) still in the group list
+			biggestGroupSize := groupList.biggestGroupSize()
+			for i := range primaryGroups {
+
```

Comment:
Unsure what's going on here.
why do we need to check all these interactions and add them to the `groupList`?
why are these groups not part of the original `groupCandidates`?

Reply:
Intersections may be not the ones that were part of the original groups.
Assuming we have the following original groups
(1, 2, 3, 4,5)  (1,2,3,4) (1,2,3,5) then only the intersection (1,2,3) is common to all original groups.